File: android/app/src/main/kotlin/com/example/mobile_banking_app/DynamicRuleParser.kt
package com.example.mobile_banking_app

import java.text.NumberFormat
import java.util.Locale

object DynamicRuleParser {

    private fun formatEtb(amount: Double): String {
        val formatter = NumberFormat.getNumberInstance(Locale.US).apply {
            minimumFractionDigits = 2
            maximumFractionDigits = 2
        }
        return "ETB ${formatter.format(amount)}"
    }

    fun parse(bank: NativeBankDefinition, body: String): SmsBroadcastReceiver.NativeParsedSms? {
        if (body.isBlank()) return null
        val singleLine = body.replace("\n", " ").replace("\r", " ")
        val lowerMsg = singleLine.lowercase()

        // 1. Security filter
        if (!bank.securityFilterRegex.isNullOrBlank()) {
            try {
                if (Regex(bank.securityFilterRegex).containsMatchIn(singleLine)) {
                    return null
                }
            } catch (_: Exception) {}
        }

        // 2. Ignore filter
        if (!bank.ignoreFilterRegex.isNullOrBlank()) {
            try {
                if (Regex(bank.ignoreFilterRegex).containsMatchIn(singleLine)) {
                    return null
                }
            } catch (_: Exception) {}
        }

        // 3. Patterns evaluation
        for (pattern in bank.patterns) {
            if (pattern.triggerKeywords.isNotEmpty()) {
                val allMatch = pattern.triggerKeywords.all { lowerMsg.contains(it) }
                if (!allMatch) continue
            }

            if (pattern.regex.isBlank()) continue

            try {
                val regExp = Regex(pattern.regex)
                val match = regExp.find(singleLine) ?: continue

                // Extract Amount
                var amount = 0.0
                if (pattern.amountGroup > 0 && pattern.amountGroup < match.groupValues.size) {
                    val rawAmt = match.groupValues[pattern.amountGroup].replace(",", "").trim()
                    val cleanAmt = if (rawAmt.endsWith(".")) rawAmt.substring(0, rawAmt.length - 1) else rawAmt
                    amount = cleanAmt.toDoubleOrNull() ?: 0.0
                }
                if (amount <= 0.0) continue

                // Extract Counterparty
                var counterparty = pattern.counterpartyDefault ?: ""
                if (pattern.counterpartyGroup > 0 && pattern.counterpartyGroup < match.groupValues.size) {
                    val cp = match.groupValues[pattern.counterpartyGroup].trim()
                    if (cp.isNotEmpty()) {
                        counterparty = cp
                    }
                }
                if (counterparty.isEmpty()) {
                    counterparty = bank.bankName
                }

                // Extract Balance
                var balance = 0.0
                if (pattern.balanceGroup != null && pattern.balanceGroup > 0 && pattern.balanceGroup < match.groupValues.size) {
                    val rawBal = match.groupValues[pattern.balanceGroup].replace(",", "").trim()
                    val cleanBal = if (rawBal.endsWith(".")) rawBal.substring(0, rawBal.length - 1) else rawBal
                    balance = cleanBal.toDoubleOrNull() ?: 0.0
                }

                // Extract Reference ID
                var txRef: String? = null
                if (pattern.idGroup != null && pattern.idGroup > 0 && pattern.idGroup < match.groupValues.size) {
                    val rawId = match.groupValues[pattern.idGroup].trim()
                    if (rawId.isNotEmpty()) {
                        txRef = rawId
                    }
                }

                val isDebit = pattern.type.equals("expense", ignoreCase = true)
                val isLocked = pattern.patternType.equals("telebirrSanduq", ignoreCase = true) ||
                               pattern.patternType.equals("internalTransfer", ignoreCase = true)
                val lockedReason = if (pattern.patternType.equals("telebirrAirtime", ignoreCase = true)) {
                    "Airtime"
                } else if (pattern.patternType.equals("telebirrPackage", ignoreCase = true)) {
                    "Package"
                } else if (pattern.patternType.equals("telebirrSanduq", ignoreCase = true)) {
                    "Sanduq"
                } else null

                return SmsBroadcastReceiver.NativeParsedSms(
                    bankName = bank.bankName,
                    amount = amount,
                    formattedAmount = formatEtb(amount),
                    isDebit = isDebit,
                    counterparty = counterparty,
                    directionHeader = if (isDebit) "To: $counterparty" else "From: $counterparty",
                    title = if (isDebit) "Expense" else "Income",
                    isLocked = isLocked,
                    lockedReasonName = lockedReason,
                    txReference = txRef,
                    totalBalance = balance
                )
            } catch (_: Exception) {
                continue
            }
        }

        return null
    }
}

File: lib/services/dynamic_rule_parser.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:intl/intl.dart';
import '../models/bank_definition.dart';
import '../models/parsed_sms_result.dart';

/// Pure-fact SMS parser driven by declarative [BankDefinition] rules.
///
/// Strictly conforms to Layer 1 architecture:
/// - Extracts ONLY the 8 objective facts in [ParsedSmsResult]
/// - Zero database, UI, or reason logic
class DynamicRuleParser {
  DynamicRuleParser._();

  static RegExp _safeRegExp(String rawPattern, {bool caseSensitive = false}) {
    String p = rawPattern;
    bool cs = caseSensitive;
    if (p.startsWith('(?i)')) {
      p = p.substring(4);
      cs = false;
    }
    return RegExp(p, caseSensitive: cs);
  }

  /// Parses an SMS [message] using the rules defined in [bank].
  ///
  /// Returns a valid [ParsedSmsResult] if a pattern matches, or null if no
  /// pattern matches or the message matches a security/ignore filter.
  static ParsedSmsResult? parse(
    BankDefinition bank,
    String message,
    DateTime fallbackDate,
  ) {
    if (message.isEmpty) return null;

    final singleLine = message.replaceAll('\n', ' ').replaceAll('\r', ' ');
    final lowerMsg = singleLine.toLowerCase();

    // 1. Security / OTP noise filter
    if (bank.parsing.securityFilterRegex != null &&
        bank.parsing.securityFilterRegex!.isNotEmpty) {
      try {
        final secRegex = _safeRegExp(bank.parsing.securityFilterRegex!);
        if (secRegex.hasMatch(singleLine)) {
          return null;
        }
      } catch (_) {}
    }

    // 2. Ignore / Promo noise filter
    if (bank.parsing.ignoreFilterRegex != null &&
        bank.parsing.ignoreFilterRegex!.isNotEmpty) {
      try {
        final ignRegex = _safeRegExp(bank.parsing.ignoreFilterRegex!);
        if (ignRegex.hasMatch(singleLine)) {
          return null;
        }
      } catch (_) {}
    }

    // 3. Evaluate Patterns in declaration order
    for (final pattern in bank.parsing.patterns) {
      // Fast check: All trigger keywords must be present
      if (pattern.triggerKeywords.isNotEmpty) {
        bool allPresent = true;
        for (final kw in pattern.triggerKeywords) {
          if (!lowerMsg.contains(kw)) {
            allPresent = false;
            break;
          }
        }
        if (!allPresent) continue;
      }

      if (pattern.regex.isEmpty) continue;

      try {
        final regExp = _safeRegExp(pattern.regex);
        final match = regExp.firstMatch(singleLine);
        if (match == null) continue;

        // Extract Amount
        double amount = 0.0;
        if (pattern.amountGroup > 0 && pattern.amountGroup <= match.groupCount) {
          final amtStr = match.group(pattern.amountGroup)?.replaceAll(',', '').trim() ?? '0';
          final cleanAmt = amtStr.endsWith('.') ? amtStr.substring(0, amtStr.length - 1) : amtStr;
          amount = double.tryParse(cleanAmt) ?? 0.0;
        }
        if (amount <= 0.0) continue;

        // Extract Counterparty
        String counterparty = pattern.counterpartyDefault ?? '';
        if (pattern.counterpartyGroup > 0 && pattern.counterpartyGroup <= match.groupCount) {
          final cp = match.group(pattern.counterpartyGroup)?.trim();
          if (cp != null && cp.isNotEmpty) {
            counterparty = cp;
          }
        }
        if (counterparty.isEmpty) {
          counterparty = bank.bankName;
        }

        // Extract Balance
        double balance = 0.0;
        if (pattern.balanceGroup != null &&
            pattern.balanceGroup! > 0 &&
            pattern.balanceGroup! <= match.groupCount) {
          final balStr = match.group(pattern.balanceGroup!)?.replaceAll(',', '').trim() ?? '0';
          final cleanBal = balStr.endsWith('.') ? balStr.substring(0, balStr.length - 1) : balStr;
          balance = double.tryParse(cleanBal) ?? 0.0;
        }

        // Extract Transaction Reference ID
        String? id;
        if (pattern.idGroup != null &&
            pattern.idGroup! > 0 &&
            pattern.idGroup! <= match.groupCount) {
          final rawId = match.group(pattern.idGroup!)?.trim();
          if (rawId != null && rawId.isNotEmpty) {
            id = rawId;
          }
        }

        // Extract Date if present
        DateTime txDate = fallbackDate;
        if (pattern.dateGroup != null &&
            pattern.dateGroup! > 0 &&
            pattern.dateGroup! <= match.groupCount &&
            pattern.dateFormat != null) {
          final dateStr = match.group(pattern.dateGroup!)?.trim();
          if (dateStr != null && dateStr.isNotEmpty) {
            try {
              final parsed = DateFormat(pattern.dateFormat!).parse(dateStr);
              txDate = DateTime(
                parsed.year < 2000 ? fallbackDate.year : parsed.year,
                parsed.month,
                parsed.day,
                parsed.hour,
                parsed.minute,
                parsed.second,
              );
            } catch (_) {}
          }
        }

        // Generate synthetic reference ID if absent
        if (id == null || id.isEmpty) {
          final content = '${bank.bankName}_${amount}_${pattern.type}_${txDate.millisecondsSinceEpoch}_$counterparty';
          id = 'SYN_${sha256.convert(utf8.encode(content)).toString().substring(0, 16)}';
        }

        return ParsedSmsResult(
          id: id,
          bankName: bank.bankName,
          amount: amount,
          type: pattern.type,
          date: txDate,
          counterparty: counterparty,
          totalBalance: balance,
          rawMessage: message,
          patternType: pattern.resolvedPatternType,
        );
      } catch (_) {
        // Safe regex error handling; fall through to next pattern
        continue;
      }
    }

    return null;
  }
}
